internal/keepers/service.go

```diff
@@ -217,76 +218,21 @@ func (s *onDemandUpkeepService) parallelCheck(ctx context.Context, keys []types.
 		return samples.Values(), nil
 	}
 
-	var (
-		cacheHits      int
-		wg             sync.WaitGroup
-		wResults       workerResults
-		keysToSendLock sync.Mutex
-		keysToSend     = make([]types.UpkeepKey, 0, len(keys))
-	)
-
-	// start the results listener
-	// if the context provided is cancelled, this listener shouldn't terminate
-	// until all results have been collected. each worker is also passed this
-	// function's context and will terminate when that context is cancelled
-	// resulting in multiple errors being collected by this listener.
-	done := make(chan struct{})
-	go s.aggregateWorkerResults(&wg, &wResults, samples, done)
+	var wResults workerResults
 
 	// go through keys and check the cache first
 	// if an item doesn't exist on the cache, send the items to the worker threads
-	for _, key := range keys {
-		wg.Add(1)
-		go func(key types.UpkeepKey) {
-			defer wg.Done()
-
-			// no RPC lookups need to be done if a result has already been cached
-			result, cached := s.cache.Get(key.String())
-			if cached {
-				cacheHits++
-				if result.State == types.Eligible {
-					samples.Append(result)
-				}
-				return
-			}
-
-			// Add key to the slice that is going to be sent to the worker queue
-			keysToSendLock.Lock()
-			keysToSend = append(keysToSend, key)
-			keysToSendLock.Unlock()
-		}(key)
-	}
-	wg.Wait()
+	filteredKeys, cacheHits := filterFromCache(keys, s.cache, samples)
 
 	// Create batches from the given keys.
 	// Max keyBatchSize items in the batch.
-	keysBatches := createBatches(keysToSend, maxWorkersBatchSize)
-	for _, batch := range keysBatches {
-		// for every job added to the worker queue, add to the wait group
-		// all jobs should complete before completing the function
-		s.logger.Printf("attempting to send keys to worker group")
-		wg.Add(1)
-		if err := s.workers.Do(ctx, makeWorkerFunc(ctx, s.logger, s.registry, batch)); err != nil {
-			if !errors.Is(err, ErrContextCancelled) {
-				// the worker process has probably stopped so the function
-				// should terminate with an error
-				close(done)
-				return nil, fmt.Errorf("%w: failed to add upkeep check to worker queue", err)
-			}
-
-			// if the context is cancelled before the work can be
-			// finished, stop adding work and allow existing to finish.
-			// cancelling this context will cancel all waiting worker
-			// functions and have them report immediately. this will
-			// result in a lot of errors in the results collector.
-			s.logger.Printf("context cancelled while attempting to add to queue")
-			wg.Done()
-		}
-	}
-
-	s.logger.Printf("waiting for results to be read")
-	wg.Wait()
-	close(done)
+	pkgutil.RunJobs(
+		ctx,
+		s.workers,
+		createBatches(filteredKeys, maxWorkersBatchSize),
+		wrapWorkerFunc(s.logger, s.registry),
+		wrapAggregate(s.logger, s.cache, &wResults, samples),
```

Comment:
Would it make sense to move both these wrapper functions into the onDemandUpkeepService, so they both have pointer receivers? That way you can cut down on what params you're sending in?

Reply:
So you'd end up with:

s.wrapWorkerFunc()

And

s.wrapAggregate(&wResults, samples)

Reply:
Makes sense. I was going for easier unit testing, but this approach connects it to the service instead which is where it belongs.